{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paper revision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load annotation feature\n",
    "\n",
    "(As we have multiple gene isoforms, we will consider only the .1 one)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import yaml\n",
    "from IPython.display import clear_output, display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation = pd.read_csv(\n",
    "    \"~/databases/vitis/V2.1.gff3\",\n",
    "    sep=\"\\t\",\n",
    "    header=None\n",
    "    )\n",
    "\n",
    "# annotation.head(n=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List of features to test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_size = {\n",
    "    \"cds\": 0,\n",
    "    \"downstream\": 0,\n",
    "    \"exon_length\": 0,\n",
    "    \"exon_number\": 0,\n",
    "    \"gene_length\": 0,\n",
    "    \"gene_number\": 0,\n",
    "    \"intergenic\": 0,\n",
    "    \"intron\": 0,\n",
    "    \"splicing_site\": 0,\n",
    "    \"start_stop\": 0,\n",
    "    \"upstream\": 0,\n",
    "    \"utr_3\": 0,\n",
    "    \"utr_5\": 0\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember, we will only consider the 1st isoform!\n",
    "for index, row in annotation[(annotation[2] == \"CDS\") & (annotation[8].str.contains(\".1.cds\"))].iterrows():\n",
    "    feature_size[\"cds\"] += row[4] - row[3]\n",
    "\n",
    "for index, row in annotation[(annotation[2] == \"exon\") & (annotation[8].str.contains(\".1.exon\"))].iterrows():\n",
    "    feature_size[\"exon_length\"] += row[4] - row[3]\n",
    "\n",
    "# Genes does not consider isoforms\n",
    "for index, row in annotation[(annotation[2] == \"gene\")].iterrows():\n",
    "    feature_size[\"gene_length\"] += row[4] - row[3]\n",
    "\n",
    "for index, row in annotation[(annotation[2] == \"three_prime_utr\") & (annotation[8].str.contains(\".1.utr3p\"))].iterrows():\n",
    "    feature_size[\"utr_3\"] += row[4] - row[3]\n",
    "\n",
    "for index, row in annotation[(annotation[2] == \"five_prime_utr\") & (annotation[8].str.contains(\".1.utr5p\"))].iterrows():\n",
    "    feature_size[\"utr_5\"] += row[4] - row[3]\n",
    "\n",
    "# print(yaml.dump(feature_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remember, we will only consider the 1st isoform!\n",
    "feature_size[\"exon_number\"] = len(annotation[(annotation[2] == \"exon\") & (annotation[8].str.contains(\".1.exon\"))])\n",
    "\n",
    "# Genes does not consider isoform\n",
    "feature_size[\"gene_number\"] = len(annotation[(annotation[2] == \"gene\")])\n",
    "\n",
    "# print(yaml.dump(feature_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_size[\"intron\"] = feature_size[\"gene_length\"] - feature_size[\"exon_length\"]\n",
    "\n",
    "junctions = (feature_size[\"exon_number\"] - feature_size[\"gene_number\"]) * 2\n",
    "feature_size[\"splicing_site\"] = junctions * 2\n",
    "\n",
    "feature_size[\"start_stop\"] = feature_size[\"gene_number\"] * 3\n",
    "\n",
    "# print(yaml.dump(feature_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset genes\n",
    "genes = annotation[annotation[2] == \"gene\"].copy()\n",
    "genes[\"upstream_start\"] = None\n",
    "genes[\"upstream_end\"] = None\n",
    "genes[\"downstream_start\"] = None\n",
    "genes[\"downstream_end\"] = None\n",
    "genes[\"domain_start\"] = None\n",
    "genes[\"domain_end\"] = None\n",
    "\n",
    "# Loop over genes to infer upstream and downstream\n",
    "for index, row in genes.iterrows():\n",
    "    if row[6] == \"+\":\n",
    "        # Upstream is before and downstream is after\n",
    "        gene_start = int(row[3])\n",
    "        gene_end = int(row[4])\n",
    "        \n",
    "        gene_upstream = gene_start - 5000\n",
    "        gene_downstream = gene_end + 5000\n",
    "        \n",
    "    else:\n",
    "        # Upstream is after and downstream is before\n",
    "        gene_start = int(row[4])\n",
    "        gene_end = int(row[3])\n",
    "        \n",
    "        gene_upstream = gene_start + 5000\n",
    "        gene_downstream = gene_start - 5000\n",
    "    \n",
    "    genes.at[index, \"upstream_start\"] = min(gene_upstream, gene_start)\n",
    "    genes.at[index, \"upstream_end\"] = max(gene_upstream, gene_start)\n",
    "    genes.at[index, \"downstream_start\"] = min(gene_downstream, gene_end)\n",
    "    genes.at[index, \"downstream_end\"] = max(gene_downstream, gene_end)\n",
    "    genes.at[index, \"domain_start\"] = min(gene_upstream, gene_downstream)\n",
    "    genes.at[index, \"domain_end\"] = max(gene_upstream, gene_downstream)\n",
    "\n",
    "# genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "820939/820939 (100.0%)\n"
     ]
    }
   ],
   "source": [
    "idx_max = genes.index.max()\n",
    "\n",
    "# Loop over genes to infer Up/Down-stream\n",
    "for index, row in genes.iterrows():\n",
    "    clear_output(wait=True)\n",
    "    print(f\"{index}/{idx_max} ({round(index/idx_max*100,1)}%)\")\n",
    "    # If no other gene touch in this gene's domain we may considere the entire upstream and downstream regions without problems\n",
    "    genes_touching = genes[(genes[0] == row[0])\n",
    "                           & (\n",
    "                               (\n",
    "                                   (genes[\"domain_start\"] >= row[\"domain_start\"])\n",
    "                                   & (genes[\"domain_start\"] <= row[\"domain_end\"])\n",
    "                               )\n",
    "                               | (\n",
    "                                   (genes[\"domain_end\"] >= row[\"domain_start\"])\n",
    "                                   & (genes[\"domain_end\"] <= row[\"domain_end\"])\n",
    "                               )\n",
    "                           )\n",
    "                           & (genes[8] != row[8])\n",
    "                          ].copy()\n",
    "    \n",
    "    if len(genes_touching.index) == 0:\n",
    "        feature_size[\"upstream\"] += 5000\n",
    "        feature_size[\"downstream\"] += 5000\n",
    "        continue\n",
    "    \n",
    "    \n",
    "    \n",
    "    ####################\n",
    "    ### UPSTREAM ONY ###\n",
    "    ####################\n",
    "    \n",
    "    # Verify if upstream touch other genes\n",
    "    genes_touching_upstream = genes_touching[(\n",
    "        ((genes_touching[\"domain_start\"] >= row[\"upstream_start\"]) & (genes_touching[\"domain_start\"] <= row[\"upstream_end\"]))\n",
    "        | ((genes_touching[\"domain_end\"] >= row[\"upstream_start\"]) & (genes_touching[\"domain_end\"] <= row[\"upstream_end\"]))\n",
    "        )].copy()\n",
    "    \n",
    "    if len(genes_touching_upstream.index) == 0:\n",
    "        feature_size[\"upstream\"] += 5000\n",
    "        # We can't continue right here, beacause we still need to evaluate downstream\n",
    "    else:\n",
    "        row_upstream_start = row[\"upstream_start\"]\n",
    "        row_upstream_end = row[\"upstream_end\"]\n",
    "        \n",
    "        # Remove portion associated to genes\n",
    "        genes_touching_upstream_gene = genes_touching_upstream[(\n",
    "            ((genes_touching_upstream[3] >= row[\"upstream_start\"]) & (genes_touching_upstream[3] <= row[\"upstream_end\"]))\n",
    "            | ((genes_touching_upstream[4] >= row[\"upstream_start\"]) & (genes_touching_upstream[4] <= row[\"upstream_end\"]))\n",
    "            )].copy()\n",
    "        \n",
    "        for second_index, second_row in genes_touching_upstream_gene.iterrows():\n",
    "            if row[6] == \"+\":\n",
    "                row_upstream_start = max(second_row[4], row_upstream_start)\n",
    "            else:\n",
    "                row_upstream_end = min(second_row[3], row_upstream_end)\n",
    "        \n",
    "        upstream_count = row_upstream_end - row_upstream_start\n",
    "        if upstream_count < 0:\n",
    "            upstream_count = 0\n",
    "        \n",
    "        else:\n",
    "            # Remove portion associated with other upstreams\n",
    "            # We will count those upstreams as half and so, when both genes are analysed, each place will sum 1!\n",
    "            \n",
    "            genes_touching_upstream_upstream = genes_touching_upstream[(\n",
    "                ((genes_touching_upstream[\"upstream_start\"] >= row_upstream_start) & (genes_touching_upstream[\"upstream_start\"] <= row_upstream_end))\n",
    "                | ((genes_touching_upstream[\"upstream_end\"] >= row_upstream_start) & (genes_touching_upstream[\"upstream_end\"] <= row_upstream_end))\n",
    "                )].copy()\n",
    "            \n",
    "            gene_upstream_range = set(range(row_upstream_start, row_upstream_end))\n",
    "            for second_index, second_row in genes_touching_upstream_upstream.iterrows():\n",
    "                second_upstream_range = set(range(second_row[\"upstream_start\"], second_row[\"upstream_end\"]))\n",
    "                upstream_intercection = len(gene_upstream_range.intersection(second_upstream_range))\n",
    "                upstream_count = upstream_count - upstream_intercection/2\n",
    "        \n",
    "        feature_size[\"upstream\"] += upstream_count\n",
    "        \n",
    "    \n",
    "    ######################\n",
    "    ### DOWNSTREAM ONY ###\n",
    "    ######################\n",
    "    \n",
    "    # Verify if downstream touch other genes\n",
    "    genes_touching_downstream = genes_touching[(\n",
    "        ((genes_touching[\"domain_start\"] >= row[\"downstream_start\"]) & (genes_touching[\"domain_start\"] <= row[\"downstream_end\"]))\n",
    "        | ((genes_touching[\"domain_end\"] >= row[\"downstream_start\"]) & (genes_touching[\"domain_end\"] <= row[\"downstream_end\"]))\n",
    "        )].copy()\n",
    "    \n",
    "    if len(genes_touching_downstream.index) == 0:\n",
    "        feature_size[\"downstream\"] += 5000\n",
    "        # We could continue right now, we we will not to keep consistency to previous code\n",
    "    else:\n",
    "        row_downstream_start = row[\"downstream_start\"]\n",
    "        row_downstream_end = row[\"downstream_end\"]\n",
    "        \n",
    "        # Remove portion associated to genes\n",
    "        genes_touching_downstream_gene = genes_touching_downstream[(\n",
    "            ((genes_touching_downstream[3] >= row[\"downstream_start\"]) & (genes_touching_downstream[3] <= row[\"downstream_end\"]))\n",
    "            | ((genes_touching_downstream[4] >= row[\"downstream_start\"]) & (genes_touching_downstream[4] <= row[\"downstream_end\"]))\n",
    "            )].copy()\n",
    "        \n",
    "        for second_index, second_row in genes_touching_downstream_gene.iterrows():\n",
    "            if row[6] == \"+\":\n",
    "                row_downstream_end = min(second_row[3], row_downstream_end)\n",
    "            else:\n",
    "                row_downstream_start = max(second_row[4], row_downstream_start)\n",
    "        \n",
    "        downstream_count = row_downstream_end - row_downstream_start\n",
    "        if downstream_count < 0:\n",
    "            downstream_count = 0\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            # Remove portion associated with upstreams\n",
    "            # Upstream has priority\n",
    "            genes_touching_downstream_upstream = genes_touching_downstream[(\n",
    "                ((genes_touching_downstream[\"upstream_start\"] >= row_downstream_start) & (genes_touching_downstream[\"upstream_start\"] <= row_downstream_end))\n",
    "                | ((genes_touching_downstream[\"upstream_end\"] >= row_downstream_start) & (genes_touching_downstream[\"upstream_end\"] <= row_downstream_end))\n",
    "                )].copy()\n",
    "\n",
    "            for second_index, second_row in genes_touching_downstream_upstream.iterrows():\n",
    "                if row[6] == \"+\":\n",
    "                    row_downstream_end = min(second_row[\"upstream_start\"], row_downstream_end)\n",
    "                else:\n",
    "                    row_downstream_start = max(second_row[\"upstream_end\"], row_downstream_start)\n",
    "\n",
    "            downstream_count = row_downstream_end - row_downstream_start\n",
    "            if downstream_count < 0:\n",
    "                downstream_count = 0\n",
    "\n",
    "            else:\n",
    "\n",
    "                # Evaluate other downstreams!            \n",
    "                # We will count those upstreams as half and so, when both genes are analysed, each place will sum 1!\n",
    "\n",
    "                genes_touching_downstream_downstream = genes_touching_downstream[(\n",
    "                    ((genes_touching_downstream[\"downstream_start\"] >= row_downstream_start) & (genes_touching_downstream[\"downstream_start\"] <= row_downstream_end))\n",
    "                    | ((genes_touching_downstream[\"downstream_end\"] >= row_downstream_start) & (genes_touching_downstream[\"downstream_end\"] <= row_downstream_end))\n",
    "                    )].copy()\n",
    "\n",
    "                gene_downstream_range = set(range(row_downstream_start, row_downstream_end))\n",
    "                for second_index, second_row in genes_touching_downstream_downstream.iterrows():\n",
    "                    second_downstream_range = set(range(second_row[\"downstream_start\"], second_row[\"downstream_end\"]))\n",
    "                    downstream_intercection = len(gene_downstream_range.intersection(second_downstream_range))\n",
    "                    downstream_count = downstream_count - downstream_intercection/2\n",
    "\n",
    "        feature_size[\"downstream\"] += downstream_count\n",
    "        \n",
    "# print(yaml.dump(feature_size))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's calculate intergenics!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_len = pd.read_csv(\"chr_len.tsv\",\n",
    "                      sep=\"\\t\",\n",
    "                      header=0,\n",
    "                      index_col=0\n",
    "                     )\n",
    "\n",
    "# chr_len.head(n=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_genome_length = chr_len[\"Length\"].sum()\n",
    "feature_size[\"intergenic\"] = int(total_genome_length - (feature_size[\"upstream\"] + feature_size[\"downstream\"] + feature_size[\"gene_length\"]))\n",
    "# feature_size[\"intergenic\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('feature_size.yaml', 'w') as output_file:\n",
    "    yaml.dump(feature_size, output_file)\n",
    "\n",
    "# print(yaml.dump(feature_size))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
